ukvm/ukvm_module_net.c

```diff
@@ -178,8 +178,10 @@ static void hypercall_netwrite(struct ukvm_hv *hv, ukvm_gpa_t gpa)
     int ret;
 
     ret = write(netfd, UKVM_CHECKED_GPA_P(hv, wr->data, wr->len), wr->len);
-    assert(wr->len == ret);
-    wr->ret = 0;
+    if (wr->len == ret)
+        wr->ret = 0;
+    else
+        wr->ret = -1;
 }
```

Comment:
Thanks for the clarification. I was not aware of (1) and read too much into the error checking code in test_ping_serve.c.

In this case, failing hard with an assertion in the write path is definitely better and I agree with your solution for (2).